src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md

```diff
@@ -0,0 +1,177 @@
+render phase. Upon each new frame or new event, `MouseTracker` compares the
+annotations hovered by the mouse pointer before and after the change, then
+dispatches callbacks accordingly. `MouseTracker` also tracks which
+annotations are on the screen by requring the annotation's owners to attach
+or detach the annotation promptly.
+
+However, since the `MouseTrackerAnnotation` is a constant class, its properties
+can't be changed without replacing the annotation as a whole, which triggers
+the exit and enter callback if it's hovered by a mouse pointer.
+
+This issue is a blocker to the plan of supporting mouse cursor, since it’s
+very likely that the cursor assigned to a region needs to be changed.
+
+
+## Description of change
+
+Class `MouseTrackerAnnotation` now requires a new parameter `key`:
+
+```diff
+ class MouseTrackerAnnotation extends Diagnosticable {
+   const MouseTrackerAnnotation({
+     this.onEnter,
+     this.onHover,
+     this.onExit,
+-  });
++    @required this.key,
++  }) : assert(key != null);
+
++  final LocalKey key;
+
+   ...
+ }
+```
+
+Three of class `MouseTracker`'s methods, `attachAnnotation`,
+`detachAnnotation`, and `isAnnotationAttached`, instead of taking a
+single parameter of `MouseTrackerAnnotation`, will now take a
+single parameter `LocalKey`:
+
+```diff
+ class MouseTracker extends ChangeNotifier {
+   ...
+
+   @visibleForTesting
+-  bool isAnnotationAttached(MouseTrackerAnnotation annotation);
++  bool isAnnotationAttached(LocalKey annotationKey);
+
+-  void attachAnnotation(MouseTrackerAnnotation annotation);
++  void attachAnnotation(LocalKey annotationKey);
+
+-  void detachAnnotation(MouseTrackerAnnotation annotation);
++  void detachAnnotation(LocalKey annotationKey);
+ }
+```
+
+## Migration guide
+
+There are 2 ways of migrating affected code. The first way requires little
+work, while the second is much simpler when the annotation's properties
+might change.
+
+### The straightforward way
+The most straightforward way is by providing
+`MouseTrackerAnnotation`'s constructor with `key: UniqueKey()`, which is
+sufficient because the annotation only needs to be equal to itself, since the
+annotation was not designed to be mutable. Then change the calls to the
+affected methods to using the annotation's key. For example:
+
+```diff
+ class MyRenderObject extends RenderObject {
+   final MouseHoverAnnotation _hoverAnnotation = MouseTrackerAnnotation(
++    key: UniqueKey(),
+     onEnter: _handleEnter,
+     onExit: _handleExit,
+   );
+
+   @override
+   void attach(PipelineOwner owner) {
+     super.attach(owner);
+-    RendererBinding.instance.mouseTracker.attachAnnotation(_hoverAnnotation);
++    RendererBinding.instance.mouseTracker.attachAnnotation(_hoverAnnotation.key);
+   }
+
+   @override
+   void detach() {
+-    RendererBinding.instance.mouseTracker.detachAnnotation(_hoverAnnotation);
++    RendererBinding.instance.mouseTracker.detachAnnotation(_hoverAnnotation.key);
+     super.detach();
```

Comment:
why does the render object do this instead of the layer tree? seems weird that you have to independently tell the layer tree and the mouse tracker about the annotations.

Reply:
There are multiple reasons:
- Annotation attachment is used to indicate widget presence. Generally, an annotation should be detached if the owner widget is unmounted, for the reason described in https://github.com/flutter/flutter/pull/44631 "The change to onExit". Therefore I think it's better to handle it in the render object (or should it be the element?).
- Historical reason. Attaching and detaching annotations used to have immediate effects (triggering callbacks) and we had to do it during Element.activate and deactivate. This was no longer needed, so now we only need to do it in attach and detach, but we haven't fully moved it to layers yet.
- Future plan. We're planning to experiment a change that split annotations from the layer tree to fix issues such as https://github.com/flutter/flutter/issues/41194. Therefore we'd like to keep it separated from the layer tree (at least for now).

Reply:
It'll still have to be in the layer tree though right? Just done differently than today (many nodes dealing with one layer rather than one layer per node).

This just seems like an unnecessarily awkward API right now.

Anyway, the point of my questions is that you should explain this in the breaking change document. Especially if we're planning on changing this again, we should warn people that we're not done changing this API.